### DISCUSSION TITLE
Cool Prop online and cool prop dll

### DISCUSSION THREAD

**Discussion by gehadrabee24** [CoolProp/CoolProp#2361] [Q&A]
I'm trying to use cool prop function in trnsys 18 to get enthalpies of air at low temperature, what i found that the online cool prop do not get the same numbers at built in function in trnsys 18, I tried too many trials and I recognized that the function did not recognize pressure change, T1=82 K and T2= 301 K and the pressure changes as following data, these values is actually correct at atmospheric pressure only (when i tried to p =0.1 mpa, i get floating point error), I checked it by the online cool prop, also if the function works only on atmospheric pressure,

delta enthalpy               Enthalpy 1                      Enthalpy 2            Pressure in Mpa
220.597156                   207.085887                 427.683043               10.000000

220.62198                     207.05882                  427.680                    1000.00

220.5969                        207.0861                    427.6830                  1.00

here is the code of the type i made: 

nRef = 729! !Air

prop(1) = -192
prop(2) = 1
Call CoolProp_Fluid_Properties(prop,nRef,12,iFlag)
If (ErrorFound()) Return
enthalpy_1 = prop(3)

prop(1) = 28
prop(2) = 1
Call CoolProp_Fluid_Properties(prop,nRef,12,iFlag)
If (ErrorFound()) Return
enthalpy_2 = prop(3)

delta_H= enthalpy_2-enthalpy_1
open(94, file = 'results.dat', status='old', position="append")
write(94,*) delta_H, enthalpy_1, enthalpy_2, prop(2)
close(94)

Also I contacted Trnsys Technical support and they checked the same values of mine and get the same enthalpies, then they told me to ask cool prop system why do I get different values from the online version and the DLL.

**Comment by ibell** [comment]
I don't follow what you did. Please show the calculated values and input values, each with units. None of the functions you are using are from CoolProp itself, so I have no clue what you actually did.

Also, provide a runnable example. None of the main developers are TRNSYS people so you need to help us help you.

**Comment by gehadrabee24** [comment]
inputs are T1=82 K and T2= 301 K, and pressure in Mpa outputs are enthalpy of air in kj/kg, the following results are at different pressures 
 Enthalpy 1       Enthalpy 2       Pressure in Mpa
 207.0861          427.6830          1.00
207.085887       427.683            10.0
 207.05882        427.680           1000.00
So, what I actually did, is creating new component on trnsys 18 which calculate the enthalpy of air at different pressure and temperature, by using the built in function of cool prop in trnsys, when i get these values i recognized the difference between them and online cool prop, so I ask technical support of  trnsys and i received this email from them:  
**Gehad -

I have confirmed that the TRNSYS interface correctly passes the pressure values to the CoolProp DLL and that I get the same values as you do. You will need to contact CoolProp to ask why you get different values from the online version and the DLL.**


you can check these values with the online coolprop, also you can check the calling code of coolprop in the following part as I can not send the file , not supported:
![image](https://github.com/CoolProp/CoolProp/assets/161880390/46b5c4a0-1b49-4cf5-a3a6-752df71c2951)

 Subroutine Type2832

! Object: CoolpropTrial
! Simulation Studio Model: Type2832
! 

! Author: Gehad
! Editor: 
! Date:	 March 19, 2024
! last modified: March 19, 2024
! 
! 
! *** 
! *** Model Parameters 
! *** 

! *** 
! *** Model Inputs 
! *** 
!			T_1	K [-Inf;+Inf]
!			T_2	K [-Inf;+Inf]
!			P	bar [-Inf;+Inf]

! *** 
! *** Model Outputs 
! *** 
!			delta_H	kJ/kg [-Inf;+Inf]
!			enthalpy_1	kJ/kg [-Inf;+Inf]
!			enthalpy_2	kJ/kg [-Inf;+Inf]

! *** 
! *** Model Derivatives 
! *** 

! (Comments and routine interface generated by TRNSYS Simulation Studio)
!************************************************************************

!-----------------------------------------------------------------------------------------------------------------------
! This TRNSYS component skeleton was generated from the TRNSYS studio based on the user-supplied parameters, inputs, 
! outputs, and derivatives.  The user should check the component formulation carefully and add the content to transform
! the parameters, inputs and derivatives into outputs.  Remember, outputs should be the average value over the timestep
! and not the value at the end of the timestep; although in many models these are exactly the same values.  Refer to 
! existing types for examples of using advanced features inside the model (Formats, Labels etc.)
!-----------------------------------------------------------------------------------------------------------------------


      Use TrnsysConstants
      Use TrnsysFunctions

!-----------------------------------------------------------------------------------------------------------------------

!DEC$Attributes DLLexport :: Type2832

!-----------------------------------------------------------------------------------------------------------------------
!Trnsys Declarations
      Implicit None

      Double Precision Timestep,Time
      Integer CurrentUnit,CurrentType
      Integer nRef, iFlagr
!    PARAMETERS

!    INPUTS
      DOUBLE PRECISION T_1
      DOUBLE PRECISION T_2
      DOUBLE PRECISION T_11
      DOUBLE PRECISION T_22
      DOUBLE PRECISION P_1
      DOUBLE PRECISION P
      Double Precision prop(9)
      DOUBLE PRECISION enthalpy_1
      DOUBLE PRECISION enthalpy_2
      DOUBLE PRECISION delta_H
      
      open(94, file = 'results.dat', status='old')  
	close(94)
!-----------------------------------------------------------------------------------------------------------------------

!-----------------------------------------------------------------------------------------------------------------------
!Get the Global Trnsys Simulation Variables
      Time=getSimulationTime()
      Timestep=getSimulationTimeStep()
      CurrentUnit = getCurrentUnit()
      CurrentType = getCurrentType()
!-----------------------------------------------------------------------------------------------------------------------

!-----------------------------------------------------------------------------------------------------------------------
!Set the Version Number for This Type
      If(getIsVersionSigningTime()) Then
		Call SetTypeVersion(17)
		Return
      EndIf
!-----------------------------------------------------------------------------------------------------------------------

!-----------------------------------------------------------------------------------------------------------------------
!Do Any Last Call Manipulations Here
      If(getIsLastCallofSimulation()) Then
		Return
      EndIf
!-----------------------------------------------------------------------------------------------------------------------

!-----------------------------------------------------------------------------------------------------------------------
!Perform Any "After Convergence" Manipulations That May Be Required at the End of Each Timestep
      If(getIsEndOfTimestep()) Then
		Return
      EndIf
!-----------------------------------------------------------------------------------------------------------------------

!-----------------------------------------------------------------------------------------------------------------------
!Do All of the "Very First Call of the Simulation Manipulations" Here
      If(getIsFirstCallofSimulation()) Then

		!Tell the TRNSYS Engine How This Type Works
		Call SetNumberofParameters(0)           !The number of parameters that the the model wants
		Call SetNumberofInputs(3)                   !The number of inputs that the the model wants
		Call SetNumberofDerivatives(0)         !The number of derivatives that the the model wants
		Call SetNumberofOutputs(3)                 !The number of outputs that the the model produces
		Call SetIterationMode(1)                             !An indicator for the iteration mode (default=1).  Refer to section 8.4.3.5 of the documentation for more details.
		Call SetNumberStoredVariables(0,0)                   !The number of static variables that the model wants stored in the global storage array and the number of dynamic variables that the model wants stored in the global storage array
		Call SetNumberofDiscreteControls(0)               !The number of discrete control functions set by this model (a value greater than zero requires the user to use Solver 1: Powell's method)

		Return

      EndIf
!-----------------------------------------------------------------------------------------------------------------------

!-----------------------------------------------------------------------------------------------------------------------
!Do All of the First Timestep Manipulations Here - There Are No Iterations at the Intial Time
      If (getIsStartTime()) Then


      T_1 = GetInputValue(1)
      T_2 = GetInputValue(2)
      P = GetInputValue(3)

	
   !Check the Parameters for Problems (#,ErrorType,Text)
   !Sample Code: If( PAR1 <= 0.) Call FoundBadParameter(1,'Fatal','The first parameter provided to this model is not acceptable.')

   !Set the Initial Values of the Outputs (#,Value)
		Call SetOutputValue(1, 0) ! delta_H
		Call SetOutputValue(2, 0) ! enthalpy_1
		Call SetOutputValue(3, 0) ! enthalpy_2


   !If Needed, Set the Initial Values of the Static Storage Variables (#,Value)
   !Sample Code: SetStaticArrayValue(1,0.d0)

   !If Needed, Set the Initial Values of the Dynamic Storage Variables (#,Value)
   !Sample Code: Call SetDynamicArrayValueThisIteration(1,20.d0)

   !If Needed, Set the Initial Values of the Discrete Controllers (#,Value)
   !Sample Code for Controller 1 Set to Off: Call SetDesiredDiscreteControlState(1,0) 

		Return

      EndIf
!-----------------------------------------------------------------------------------------------------------------------

!-----------------------------------------------------------------------------------------------------------------------
!ReRead the Parameters if Another Unit of This Type Has Been Called Last
      If(getIsReReadParameters()) Then
		!Read in the Values of the Parameters from the Input File

		
      EndIf
!-----------------------------------------------------------------------------------------------------------------------

!Read the Inputs
      T_1 = GetInputValue(1)
      T_2 = GetInputValue(2)
      P = GetInputValue(3)
		

	!Check the Inputs for Problems (#,ErrorType,Text)
	!Sample Code: If( IN1 <= 0.) Call FoundBadInput(1,'Fatal','The first input provided to this model is not acceptable.')
 
      If(ErrorFound()) Return
!-----------------------------------------------------------------------------------------------------------------------

!-----------------------------------------------------------------------------------------------------------------------
!    *** PERFORM ALL THE CALCULATION HERE FOR THIS MODEL. ***
!-----------------------------------------------------------------------------------------------------------------------
        nRef = 729!729!Air
        T_11=(T_1-273)
        P_1=(P/10)
        prop(1) = T_11 !K
        prop(2) = P_1    !Mpa
        Call CoolProp_Fluid_Properties(prop,nRef,12,iFlagr)
        If (ErrorFound()) Return
      enthalpy_1 = prop(3)

        
        T_22=(T_2-273)
        prop(1) = T_22 !K
        prop(2) = P_1 !Mpa
        Call CoolProp_Fluid_Properties(prop,nRef,12,iFlagr)
        If (ErrorFound()) Return
        
         enthalpy_2 = prop(3)

        delta_H= enthalpy_2-enthalpy_1
       open(94, file = 'results.dat', status='old', position="append")  
        write(94,*)  delta_H, enthalpy_1, enthalpy_2, prop(2)
      close(94)
        !print*,delta_H
      
        
	!-----------------------------------------------------------------------------------------------------------------------
	!If Needed, Get the Previous Control States if Discrete Controllers are Being Used (#)
	!Sample Code: CONTROL_LAST=getPreviousControlState(1)
	!-----------------------------------------------------------------------------------------------------------------------

	!-----------------------------------------------------------------------------------------------------------------------
	!If Needed, Get the Values from the Global Storage Array for the Static Variables (#)
	!Sample Code: STATIC1=getStaticArrayValue(1)
	!-----------------------------------------------------------------------------------------------------------------------

	!-----------------------------------------------------------------------------------------------------------------------
	!If Needed, Get the Initial Values of the Dynamic Variables from the Global Storage Array (#)
	!Sample Code: T_INITIAL_1=getDynamicArrayValueLastTimestep(1)
	!-----------------------------------------------------------------------------------------------------------------------

	!-----------------------------------------------------------------------------------------------------------------------
	!Perform All of the Calculations Here to Set the Outputs from the Model Based on the Inputs

	!Sample Code: OUT1=IN1+PAR1

	!If the model requires the solution of numerical derivatives, set these derivatives and get the current solution
	!Sample Code: T1=getNumericalSolution(1)
	!Sample Code: T2=getNumericalSolution(2)
	!Sample Code: DTDT1=3.*T2+7.*T1-15.
	!Sample Code: DTDT2=-2.*T1+11.*T2+21.
	!Sample Code: Call SetNumericalDerivative(1,DTDT1)
	!Sample Code: Call SetNumericalDerivative(2,DTDT2)

!-----------------------------------------------------------------------------------------------------------------------

!-----------------------------------------------------------------------------------------------------------------------
!Set the Outputs from this Model (#,Value)
		Call SetOutputValue(1, delta_H) ! delta_H
		Call SetOutputValue(2, enthalpy_1) ! enthalpy_1
		Call SetOutputValue(3, enthalpy_2) ! enthalpy_2

!-----------------------------------------------------------------------------------------------------------------------

!-----------------------------------------------------------------------------------------------------------------------
!If Needed, Store the Desired Disceret Control Signal Values for this Iteration (#,State)
!Sample Code:  Call SetDesiredDiscreteControlState(1,1)
!-----------------------------------------------------------------------------------------------------------------------

!-----------------------------------------------------------------------------------------------------------------------
!If Needed, Store the Final value of the Dynamic Variables in the Global Storage Array (#,Value)
!Sample Code:  Call SetDynamicArrayValueThisIteration(1,T_FINAL_1)
!-----------------------------------------------------------------------------------------------------------------------
 
      Return
      End
!-----------------------------------------------------------------------------------------------------------------------

I hope I helped you what the actual problem, the dll file of coolprop in trnsys 18 provides wrong outputs.

**Comment by ibell** [comment]
None of that is helpful for figuring out what is going on. An example like this is required:
```python
import CoolProp.CoolProp as CP 

for T in [82, 301]:
    for p_MPa in [1, 10, 1000]:
        print(T, p_MPa, CP.PropsSI('Hmass','T',T,'P',p_MPa*1e6,'Air'))
```
yielding
```
82 1 6586.769669726461
82 10 12684.331114429517
82 1000 750924.8048955632
301 1 425301.3583296095
301 10 407355.9477635275
301 1000 1097476.8132131032
```

Do you get these results? If not, have you set the reference state in your code?

**Comment by gehadrabee24** [comment]
no i do not get these results, i wrote to you above what results i get, 
![image](https://github.com/CoolProp/CoolProp/assets/161880390/d9401188-b973-4c1e-ac5c-7efc17c56817)
 
 And I do not use python, I use the fortran environment on Trnsys 18, the calling code is above and main calling statement is  

    Call CoolProp_Fluid_Properties(prop,nRef,12,iFlagr)
here is the main code form programmer guide on trnsys 
![image](https://github.com/CoolProp/CoolProp/assets/161880390/4223090a-e2be-493a-a7de-3222517b708c)

**Comment by ibell** [comment]
The coolprop online results are the same as from the Python code, as I expected. The difference is caused by TRNSYS, so they will have to debug on their side. I suspect they may be doing something like setting the reference state, but without seeing how they call CoolProp it is impossible to say. Besides, you should only ever be comparing differences in enthalpy between state points.

**Comment by gehadrabee24** [comment]
I know, If you see the first message I was writing delta Enthalpy as the first result, I'll send to them again, but do you think if it's problem of dll file as they say, can we just change this dll file, by correct one from you?

**Comment by ibell** [comment]
As I mentioned in my response, the DLL/python and the online interface are giving the same result so everything is consistent on the CoolProp side. Why TRNSYS gives a different result I have no idea. As I said, you need to contact TRNSYS to understand why their result is different.

```python
import CoolProp.CoolProp as CP 
for p_MPa in [0.1, 1, 10, 100, 1000]:
    T1 = 82 # K
    T2 = 301 # K 
    h1 = CP.PropsSI('Hmass','T',T1,'P',p_MPa*1e6,'Air')/1e3
    h2 = CP.PropsSI('Hmass','T',T2,'P',p_MPa*1e6,'Air')/1e3
    print(p_MPa, h2-h1)
```
yielding
```
0.1 222.15678488831094
1 418.7145886598831
10 394.671616649098
100 325.33639328069944
1000 346.55200831754007
```

**Comment by gehadrabee24** [comment]
Hii
I'm trying to get quality of air at 107 K and P=1 bar and the result= -1, I tried on online coolprop and the built in coolprop in TRNSYS, how to solve this bug?

**Reply by ibell** [reply]
The vapor quality is ONLY defined for the range [0,1], all other values are only for guidance, and indicate that that the state is NOT two-phase. The dew and bubble temperatures at 1 bar calculated by:
```python
import CoolProp.CoolProp as CP 

p_MPa = 0.101325
Tdew = CP.PropsSI('T','Q',1,'P',p_MPa*1e6,'Air')
Tbub = CP.PropsSI('T','Q',0,'P',p_MPa*1e6,'Air')
print(f'Tdew: {Tdew} K')
print(f'Tbub: {Tbub} K')
```
yielding
```
Tdew: 81.72003595240088 K
Tbub: 78.90295718413455 K
```

**Comment by henningjp** [comment]
@gehadrabee24 , I think there are some possible issues here:

1. Your input pressure of P = 1 bar = 0.1 MPa
    At this pressure, the saturated vapor temp is 81.609 K, so your temp (107 K) is in the Vapor phase and Q is set to -1 outside of the two-phase region.  This is the correct result from CoolProp, not a bug.
2. The bigger issue is that in your code above, pressures are getting passed to the property function in MPa, but PropsSI (CoolProp) takes pressures in Pa.  Note @ibell 's python code above multiplied the MPa values by 1e6.  Your code should be doing the same before you pass the pressure to the property function.  Your code has been making calls at 1, 10, and 100 Pa, not MPa.  This should straighten out most of your issues.

**Comment by ibell** [comment]
What are dew and bubble temps for given pressure? (Or same for pressures
for given T) Vapor quality is not in general defined for pseudo fluid eos
anyhow.

**Comment by gehadrabee24** [comment]
the same pressure (P=1.01325 bar) and i tried T=78, 82 and various temperatures and still quality =-1 when i tried to get it from the enthalpy equation : h= h_fg * x + h_f  , where h_fg = enthalpy(Air_ha,x=1,P=P) - enthalpy(Air_ha,x=0,P=P)
h_f = enthalpy(Air_ha,x=0,P=P), h=enthalpy(Air_ha,T=78,P=P) i get x in negative number, but when i tried it by EES engineering equation solver, x=0.316
EES reference if refprop as i remember
